File: docs/src/main/paradox/tutorial/networking/basic/pingpong/src/main/java/se/sics/test/Parent.java
package se.sics.test;

import se.sics.kompics.ComponentDefinition;
import se.sics.kompics.Component;
import se.sics.kompics.Init;
import se.sics.kompics.timer.Timer;
import se.sics.kompics.timer.java.JavaTimer;
import se.sics.kompics.network.Network;
import se.sics.kompics.network.netty.NettyNetwork;
import se.sics.kompics.network.netty.NettyInit;

public class Parent extends ComponentDefinition {
	public Parent(Init init) {
		Component timer = create(JavaTimer.class, Init.NONE);
		Component network = create(NettyNetwork.class, new NettyInit(init.self));
		Component pinger = create(Pinger.class, new Pinger.Init(init.self));
		Component ponger = create(Ponger.class, new Ponger.Init(init.self));
		Component pinger2 = create(Pinger.class, new Pinger.Init(init.self));


		connect(pinger.getNegative(Timer.class), timer.getPositive(Timer.class));
		connect(pinger2.getNegative(Timer.class), timer.getPositive(Timer.class));

		connect(pinger.getNegative(Network.class), network.getPositive(Network.class));
		connect(pinger2.getNegative(Network.class), network.getPositive(Network.class));
		connect(ponger.getNegative(Network.class), network.getPositive(Network.class));
	}
	public static class Init extends se.sics.kompics.Init<Parent> {
		public final TAddress self;
		public Init(TAddress self) {
			this.self = self;
		}
	}
}

File: docs/src/main/scala/sexamples/virtualnetworking/pingpongvirtual/Parent.scala
package sexamples.virtualnetworking.pingpongvirtual

import com.google.common.primitives.Ints;
import se.sics.kompics.sl._
import se.sics.kompics.Channel;
import se.sics.kompics.timer.Timer
import se.sics.kompics.timer.java.JavaTimer
import se.sics.kompics.network.Network
import se.sics.kompics.network.netty.{NettyInit, NettyNetwork}
import se.sics.kompics.network.virtual.VirtualNetworkChannel

// #pinger-parent
class PingerParent extends ComponentDefinition {

  val baseSelf = cfg.getValue[TAddress]("pingpong.pinger.addr");

  val timer = create[JavaTimer];
  val network = create[NettyNetwork](new NettyInit(baseSelf));

  val vnc = VirtualNetworkChannel.connect(network.getPositive(classOf[Network]), this.proxy);

  val num = cfg.getValue[Int]("pingpong.pinger.num");
  for (i <- 0 until num) {
    val id = Ints.toByteArray(i);
    val cbuild = this.config().modify(this.id());
    cbuild.setValue("pingpong.pinger.addr", baseSelf.copy(id = Some(id)));
    val pinger = create[Pinger](Init.none[Pinger], cbuild.finalise());

    connect[Timer](timer -> pinger);

    vnc.addConnection(id, pinger.getNegative(classOf[Network]));
  }
}
// #pinger-parent

// #ponger-parent
class PongerParent extends ComponentDefinition {

  val self = cfg.getValue[TAddress]("pingpong.ponger.addr");

  val network = create[NettyNetwork](new NettyInit(self));
  val ponger = create[Ponger];

  connect[Network](network -> ponger);
}
// #ponger-parent
